[Quant] Add quantization 2.0 document

leslie-fang-intel opened the pull request:
Fixes [#issue_2336](https://github.com/pytorch/tutorials/issues/2336)

## Description
Add the new document for Quantization 2.0 flow.

## Checklist
<!--- Make sure to add `x` to all items in the following checklist: -->
- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnessessary issues are included into this pull request.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @ZailiWang @ZhaoqiongZ @Xia-Weiwen @sekahler2 @CaoE @zhuhaozhe @Valentine233

leslie-fang-intel commented:
@jgong5 @mingfeima @Xia-Weiwen Here is the draft of documents for Quantization 2.0.

jgong5 commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+2. Define the BackendQuantizer
+--------------------------------------------------------
+
+Then we will define the skeleton of a BackendQuantizer. The annotatation methods for each operation will be
+defined later.
```
Suggest to explain the semantics of each method of the quantizer a backend needs to implement.

leslie-fang-intel replied:
Thanks for the comment. Add docstring for each method.

jgong5 commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+            observer_or_fake_quant_ctr=bias_observer_or_fake_quant_ctr
+        )
+        quantization_config = QuantizationConfig(
+            act_quantization_spec, weight_quantization_spec, bias_quantization_spec, is_qat
```
Where is `is_qat` defined?

leslie-fang-intel replied:
Thanks, remove it.

jgong5 commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+1. Define QuantizationConfig
+--------------------------------------------------------
+
+QuantizationConfig defines the data type and qscheme for activation, weight and bias. Suppose
+we want to define:
+
+-  Activation: `int8` data type, `per_tensor_affine` quantization, `HistogramObserver`
+-  Weight    : `int8` data type, `per_channel_symmetric` quantization, `PerChannelMinMaxObserver`
+-  Bias      : `float` data type, `PlaceholderObserver`
```
Can we explain how the qconfig defined here how the defined `QuantizationSpec` here are used or related to the `QuantizationSpec`  used for annotating the model.

leslie-fang-intel replied:
Add more explanation for how `QuantizationConfig` is defined and how does `QuantizationSpec` been used to create observer when annotating the model.

jgong5 commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+observe certain tensor values in the prepare step. Now, we will have a step by step
+tutorial for how to use the `QuantizationAnnotation API` to create a quantizer.
+
+1. Define QuantizationConfig
```
Before going into details, it would be helpful to explain some key concepts and how they are related and involved in the quantization flow, e.g., `QuantizationAnnotation`, `QuantizationSpec` (various kinds of spec) and `QuantizationConfig`.

leslie-fang-intel replied:
Thanks for the suggestions. I have added further explanations. Please help to take a look again.

jerryzh168 replied:
this is actually not part of the annotation API, I think it's better to put this in the example quantizer section

leslie-fang-intel replied:
Thanks for the suggestion. I think the definition of ``QuantizationConfig`` and skeleton of ``BackendQuantizer`` can introduce the basic concepts such as ``QuantizationConfig``, ``QuantizationSpec``. These concepts make user easier to understand what has been done inside the ``annotation API``. So I prefer to put it before introducing several use cases of `annotation API`. Please kindly let me know your thoughts.

jerryzh168 replied:
`QuantizationConfig` is not part of the annotation api, annotation api includes: different types of QuantizationSpec and QuantizationAnnotation object

leslie-fang-intel replied:
Got it. I think `QuantizationConfig` will still be used in the final toy example. If we remove this section, will it be confused for user to run the final toy example?

leslie-fang-intel commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+
+Suppose we are a backend developer and we wish to integrate our backend
+with PyTorch's quantization 2.0 flow. We only need to define the backend
+specific quantizer. An existing quantizer object defined for
```
Thanks for the suggestions. The high level arch diagram has been added.

leslie-fang-intel commented:
> Please make sure to add a `customcarditem` to https://github.com/pytorch/tutorials/blob/main/prototype_source/prototype_index.rst

@svekars Thanks for the suggestion. `customcarditem` has been added. But I keep seeing some preCI failures, the error message  is `ERROR: botocore 1.12.25 has requirement urllib3<1.24,>=1.20, but you'll have urllib3 2.0.2 which is incompatible.`  Do you have any suggestions?

leslie-fang-intel commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+3. Annotate common operator patterns
+--------------------------------------------------------
+
+Now we will start to define the annotatation methods inside quantizer. For common operators like `conv2d`, we can use `QuantizationSpec` to
+annotate the input, weight, bias and output.
```
Thanks, add the explanation of `get_act_qspec`, `get_weight_qspec` and `get_bias_qspec` together with `QuantizationConfig` and `QuantizationSpec` previously.

leslie-fang-intel commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,567 @@
+            self._annotate_linear(model, config)
+            for node in reversed(model.graph.nodes):
+                self._annotate_conv2d(node, config)
+                self._annotate_maxpool2d(node, config)
```
Thanks for the comment. Based on latest PRs, they can been implemented similarly now. I have unified it.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+(prototype) PyTorch Quantization 2.0 Tutorial
+==========================================
```
```suggestion
=====================================================
```

leslie-fang-intel replied:
Done

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+which uses symbolic_trace to capture the model into a graph, and then
+perform quantization transformations on top of the captured model. In a
+similar way, for Quantization 2.0 flow, we will now use the PT2 Export
+workflow to capture the model into a graph, and perform quantizations
```
```suggestion
workflow to capture the model into a graph, and perform quantization
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+perform quantization transformations on top of the captured model. In a
+similar way, for Quantization 2.0 flow, we will now use the PT2 Export
+workflow to capture the model into a graph, and perform quantizations
+transformations on top of the ATen dialect graph. This is expected to
```
```suggestion
transformations on top of the ATen dialect graph. This approach is expected to
```

leslie-fang-intel replied:
Done

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+have significantly higher model coverage, better programmability, and
+a simplified UX.
+
+Suppose we are a backend developer and we wish to integrate our backend
```
```suggestion
Imagine a backend developer who wishes to integrate a third-party backend
```

leslie-fang-intel replied:
Thanks for the suggestions and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+Suppose we are a backend developer and we wish to integrate our backend
+with PyTorch's quantization 2.0 flow. We only need to define the backend
+specific quantizer. The high level arch of quantization 2.0 with quantizer could be: 
```
```suggestion
specific quantizer. The high level architecture of quantization 2.0 with quantizer could look like this: 
```

leslie-fang-intel replied:
Thanks for the suggestions and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+Inside the Quantizer, we will use the `QuantizationAnnotation API <https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#>`__
+to convey user's intent for what quantization spec to use and how to
+observe certain tensor values in the prepare step. Now, we will have a step by step
```
```suggestion
observe certain tensor values in the prepare step. Now, we will have a step-by-step
```

leslie-fang-intel replied:
Thanks for the suggestions and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+observe certain tensor values in the prepare step. Now, we will have a step by step
+tutorial for how to use the `QuantizationAnnotation API` to create a quantizer.
+
+1. Define QuantizationConfig
```
```suggestion
1. Define ``QuantizationConfig``
```

leslie-fang-intel replied:
Thanks and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+`QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
+consists of `QuantizationSpec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L28-L66>`__
+for activation, weight and bias seperately. Each `QuantizationSpec` defines the data type, qscheme and other quantization parameters used to create the observer.
```
```suggestion
for activation, weight, and bias separately. Each `QuantizationSpec` defines the data type, ``qscheme``, and other quantization parameters used to create the observer.
```

leslie-fang-intel replied:
Thanks and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+`get_weight_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#L26>`__ and
+`get_bias_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#LL42C5-L42C19>`__
+are used to get the `QuantizationSpec` from `QuantizationConfig` for a specific node. Then corresponding observer will be created
+based on this node's `QuantizationSpec`. Suppose we want use these quantization parameters for activation, weight and bias:
```
```suggestion
based on this node's `QuantizationSpec`. For example, we want to use these quantization parameters for activation, weight, and bias:
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+-  Weight    : `int8` data type, `per_channel_symmetric` quantization, `PerChannelMinMaxObserver`
+-  Bias      : `float` data type, `PlaceholderObserver`
+
+We can define the `QuantizationConfig` as below:
```
```suggestion
We can define the ``QuantizationConfig`` as below:
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+        )
+        return quantization_config
+
+2. Define the BackendQuantizer
```
```suggestion
2. Define the ``BackendQuantizer``
```

leslie-fang-intel replied:
Thanks and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+--------------------------------------------------------
+
+Now we will start to define the annotatation methods inside quantizer. For common operators like `conv2d`, we can use `QuantizationSpec` to
+annotate the input, weight, bias and output.
```
```suggestion
annotate the input, weight, bias, and output.
```

leslie-fang-intel replied:
Thanks and changed.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+4. Annotate sharing qparams operators
+--------------------------------------------------------
+
+For operator such as `add` and `cat`, which we want the two inputs sharing
```
```suggestion
For operator such as ``add`` and ``cat``, which we want the two inputs sharing
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+--------------------------------------------------------
+
+For operator such as `add` and `cat`, which we want the two inputs sharing
+quantization parameters, we can use the `SharedQuantizationSpec` to make the two inputs
```
```suggestion
quantization parameters, we can use the ``SharedQuantizationSpec`` to make the two inputs
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+--------------------------------------------------------
+
+For operator such as `sigmoid`, which has predefined and fixed scale/zero_point,
+we can use fixed parameters for it with `FixedQParamsQuantizationSpec`.
```
```suggestion
we can use fixed parameters for it with ``FixedQParamsQuantizationSpec``.
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+6. Annotate tensor with derived quantization parameters
+--------------------------------------------------------
+
+`DerivedQuantizationSpec` is the quantization spec for the Tensors whose quantization parameters are derived from other Tensors.
```
```suggestion
``DerivedQuantizationSpec`` is the quantization spec for the Tensors whose quantization parameters are derived from other Tensors.
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+--------------------------------------------------------
+
+`DerivedQuantizationSpec` is the quantization spec for the Tensors whose quantization parameters are derived from other Tensors.
+For example, we want to define the scale, zp for bias derived from activation and weight of convolution node.
```
```suggestion
For example, we want to define the scale, ``zp`` for bias derived from activation, and weight of convolution node.
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+7. A Toy Example with Resnet18 
+--------------------------------------------------------
+
+After above annotation methods defined with `QuantizationAnnotation API`, we can now put them together for the BackendQuantizer
```
```suggestion
After above annotation methods defined with ``QuantizationAnnotation API``, we can now put them together for the ``BackendQuantizer``
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+        quantizer.set_global(operator_config)
+        m = prepare_pt2e_quantizer(m, quantizer)
+        after_prepare_result = m(*example_inputs)
+        m = convert_pt2e(m)
```
Can we please add a conclusion with that sum ups what the user has learned, what else to read/run with links to extra resources if applicable.

leslie-fang-intel replied:
Good suggestions. The conclusion and further reading links are added.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+a simplified UX.
+
+Suppose we are a backend developer and we wish to integrate our backend
+with PyTorch's quantization 2.0 flow. We only need to define the backend
```
```suggestion
with PyTorch's quantization 2.0 flow. To accomplish this, they would only need to define the backend
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+An existing quantizer object defined for QNNPack/XNNPack is here 
+`QNNPackQuantizer <https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py>`__.
+Taking QNNPackQuantizer as example, the overall Quantization 2.0 flow could be:
```
```suggestion
Taking QNNPackQuantizer as an example, the overall Quantization 2.0 flow could be:
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+.. image:: /_static/img/quantization/pytorch_quantization_2_0_diagram.png
+
+An existing quantizer object defined for QNNPack/XNNPack is here 
```
```suggestion
An existing quantizer object defined for QNNPack/XNNPack is located in 
```

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+transformations on top of the ATen dialect graph. This is expected to
+have significantly higher model coverage, better programmability, and
+a simplified UX.
+
```
I would add prerequisites here. Something like this

Prerequisites
-------------

This tutorial requires the following:
- Understanding of quantization concepts in PyTorch - link to resource (other tutorials or docs)
- Pytorch 2.0 or later
- ....

leslie-fang-intel replied:
Thanks for the suggestions. The Prerequisites are added.

svekars commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+
+Today we have `FX Graph Mode
+Quantization <https://pytorch.org/docs/stable/quantization.html#prototype-fx-graph-mode-quantization>`__
+which uses symbolic_trace to capture the model into a graph, and then
```
```suggestion
which uses ``symbolic_trace`` to capture the model into a graph, and then
```

svekars reviewed (commented):
A few editorial suggestions.

svekars reviewed (commented):
A few editorial suggestions.

jerryzh168 commented on _static/img/quantization/pytorch_quantization_2_0_diagram.png:
could you make this a ascii graph so that it can be edited later? something like the graph in https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/README.md

leslie-fang-intel replied:
Thanks for the suggestion and changed to the ascii graph.

leslie-fang-intel replied:
Looks like the ascii graph didn't render correctly as in preview: https://deploy-preview-2354--pytorch-tutorials-preview.netlify.app/prototype/quantization_in_pytorch_2_0_tutorial.html @svekars @jerryzh168 Do you have any suggestions how to make it render correctly?

leslie-fang-intel replied:
After put the ascii graph into the code block, the rendering graph looks reasonable now.

jerryzh168 commented on prototype_source/quantization_2_0_tutotial.rst:
```diff
@@ -0,0 +1,595 @@
+(prototype) PyTorch Quantization 2.0 Tutorial
```
nit: should be: Quantization in PyTorch 2.0 Export

leslie-fang-intel replied:
Thanks. The title and filename updated accordingly.

leslie-fang-intel commented:
> A few editorial suggestions.

Fixed, @svekars please help to take a look again.

leslie-fang-intel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,646 @@
+    |                    Dynamo Export                     |
+    —-------------------------------------------------------
+                                |
+                        FX Graph in CATen    QNNPackQuantizer,
```
Thanks and Changed.

leslie-fang-intel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,646 @@
+            return model
+
+        def validate(self, model: torch.fx.GraphModule) -> None:
+            """validate the annotated graph is supported by the backend"""
```
Thanks and changed.

leslie-fang-intel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,646 @@
+
+        @classmethod
+        def get_supported_operators(cls) -> List[OperatorConfig]:
+            """return the operator list which is supported by the backend"""
```
Add my understanding about these questions:

- The semantic of `get_supported_operators`.
- The semantic of `OperatorConfig`.

Hi @jerryzh168, I think we need your help here for further elaboration about `get_supported_operators` since it's a  abstractmethod in base `Quantizer` which must be implemented by BackendQuantizer.

jerryzh168 replied:
I don't think these need to be in this doc, the main goal for this doc is to introduce different types of quantization specs and how to annotate patterns

leslie-fang-intel replied:
Got it, but we still need to define this method `get_supported_operators` for final toy example since it's an abstractmethod in base Quantizer which must be implemented by BackendQuantizer. So, did you suggest we skip this section and directly keep it in the final toy example?

jerryzh168 commented:
at the high level I think we should have more explanations of different quantization spec and their use cases, you can refer to https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit# and the code https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/quantizer.py to explain in more detail about what these are and how to use them I think

we are still thinking of simplifying the pattern matching related APIs and maybe some parts of the doc will need to be updated after these changes lands. cc @kimishpatel is working on that currently.

leslie-fang-intel commented:
> at the high level I think we should have more explanations of different quantization spec and their use cases, you can refer to https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit# and the code https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/quantizer.py to explain in more detail about what these are and how to use them I think
> 
> we are still thinking of simplifying the pattern matching related APIs and maybe some parts of the doc will need to be updated after these changes lands. cc @kimishpatel is working on that currently.

Hi @jerryzh168, I have updated this document per our discussion. Please help to take a look again.

kimishpatel commented:
> at the high level I think we should have more explanations of different quantization spec and their use cases, you can refer to https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit# and the code https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/quantizer.py to explain in more detail about what these are and how to use them I think
> 
> we are still thinking of simplifying the pattern matching related APIs and maybe some parts of the doc will need to be updated after these changes lands. cc @kimishpatel is working on that currently.

I want to highlight that pattern matching is not core part of quantization really. It is implementation details of quantizer, who while we showcase examples of different ways of pattern matching, it is orthogonal to quantization.

svekars reviewed (approved):
From the publishing perspective, LGTM! Make sure to address @kimishpatel concerns.

kimishpatel commented:
> From the publishing perspective, LGTM! Make sure to address @kimishpatel concerns.

Actually I am gonna push back a bit. I havent had time to review this well yet.

kimishpatel reviewed (changes requested):
Just want to get some time to review hence "requesting change". My comments will follow.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,627 @@
+input tensors and output tensor of this operator which includes parameters of ``observer type``, ``dtype``,
+``quant_min``, and ``quant_max`` etc.
+
+::
```
I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:
1. identify conv2d pattern
2. derive the quantization spec for input and output Tensors
3. annotate the nodes in the pattern properly

you can also talk about how to annotate conv - relu pattern in this section

jerryzh168 replied:
maybe start with add / add - relu as an example here, since conv requires DerivedQuantizationSpec for bias

jerryzh168 replied:
you can follow the steps in https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#heading=h.it9h4gjr7m9g for each section that introduces a new quantization spec I think

leslie-fang-intel replied:
> I feel we do not necessarily need to show this code, but instead it will be helpful to start with a use case here, let's say, we can talk about the use case of a quantized conv2d, it should go through these feel steps:

Hi @jerryzh168, Thanks for the suggestions. I have updated the document by elaboration more about use case and steps to use each ``QuantizationSpec``.  Please help to take a look again.
One question is about ``DerivedQuantizationSpec``, I think this [document ](https://docs.google.com/document/d/1tjIsL7-uVgm_1bv_kUK7iovP6G1D5zcbzwEcmYEG2Js/edit#) didn't introduce motivation clearly. Since we generally use `PlaceholderObserver` with `QuantizationSpec` for bias node, why we have to use ``DerivedQuantizationSpec``?

jerryzh168 replied:
I see, this is for the Tensors whose quantization parameters are derived from other Tensors, for example bias, the `PlaceholderObserver` that we used previously did not express this relation correctly

leslie-fang-intel replied:
Hi @jerryzh168. Thanks for the comment. Will the [`bias_quantization_spec`](https://github.com/pytorch/pytorch/blob/a02a58d862b00a7aa410149e8cf276b8d0ca3462/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L183-L185) in `QNNPackQuantizer` updated to use `DerivedQuantizationSpec`? If so, does it mean there will be the `fp32_bias -> quant_to_int32 -> dequant -> conv` pattern instead of `fp32_bias -> conv` after `convert_pt2e`?

jerryzh168 replied:
yeah I think it will be updated eventually, but I don't know when we plan to do that yet, @kimishpatel would have more context on the timeline

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,627 @@
+2. Annotate sharing qparams operators
+--------------------------------------------------------
+
+`SharedQuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L90>`__
```
similarly I think it's better to start with the use case of what quantized add/cat expects for its input and what the output would look like for quantized add, then derive the quantization spec for input/outputs and describe how to annotate the nodes in the add/cat pattern

leslie-fang-intel replied:
Sure, changed accordingly.

leslie-fang-intel commented:
> Just want to get some time to review hence "requesting change". My comments will follow.

Hi @kimishpatel, thanks for the comment and looking forward to your further suggestions. One question is that this docathon event may end at June 13. So it will be great if we can review and modify it early.

leslie-fang-intel commented:
> > > Just want to get some time to review hence "requesting change". My comments will follow.
> > 
> > 
> > Hi @kimishpatel, thanks for the comment and looking forward to your further suggestions. One question is that this docathon event may end at June 13. So it will be great if we can review and modify it early.
> 
> My apologies for late responses. Its been bit hectic.
> 
> I am gonna leave high level comment at the moment.
> 
> We really need to aim to demonstrate:
> 
> * Quantizer.
>   
>   * what is the role of quantizer.
>   * what are its APIs
> * Annotation API
>   
>   * Role of annotation API
>   * API
> * Quantized model representation
>   
>   * This one is being actively worked upon but at the moment quantized model has decomposed q/dq ops
> 
> Example implementation of quantizer is meant for illustration only and not meant to be the only way to say pattern match or find set of nodes to annotate. So those should just go in the examples.
> 
> cc @jerryzh168 for further thoughts

Thanks for the suggestions @kimishpatel. Based on the original [issue-2336](https://github.com/pytorch/tutorials/issues/2336) we try to resolve in this docathon event and the previous discussion with @jerryzh168 in this PR, my understanding is we should only focus on `Annotation API` parts in this doc. As for the `Quantized model representation` and general design of `Quantizer`, we may include them in another document in the future.

leslie-fang-intel commented:
Just get a refreshed email, the deadline for submission of this doc for docathon is `June 11, 2023 at 5PM PST`. So I think we had better not change the scope of this document at this moment. @jerryzh168 @kimishpatel what's your suggestions?

jerryzh168 commented:
yeah I think we can focus on Annotation API, but we do need to mention Quantizer as well since that is where Annotation API is used, I think

Quantizer.
* what is the role of quantizer.
* what are its APIs

this is not too hard to explain right, is there anything missing?

for Quantized model representation, I agree that we can delay this for later

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+
+Imagine a backend developer who wishes to integrate a third-party backend
+with PyTorch's quantization 2.0 flow. To accomplish this, they would only need
+to define the backend specific quantizer. The high level architecture of
```
I think we would need to mention what quantizer is here

leslie-fang-intel replied:
Any suggestions about how to introduce quantizer API of `get_supported_operators`, as we discussed here https://github.com/pytorch/tutorials/pull/2354#discussion_r1213834401. It's not finalized yet, but it's a must have when define a `Quantizer`.

jerryzh168 replied:
I think you can skip explaining that here, please focus on the role of Quantizer and the annotate function, you can also try to have a quick summarize of https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c here when talking about the motivation of the quantizer API

leslie-fang-intel replied:
Thanks. Add the introduction of ``Quantizer`` such as why we introduce this class and its APIs.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+) input, output of the operator. Following is an example flow (with ``add``)
+of how this intent is conveyed in the quantization workflow with node annotation API.
+
+-  Step1: Identify the original floating point ``add`` node in the FX graph. There are
```
nit: space between Step and 1. same for other things

jerryzh168 replied:
also we should be identifying a pattern (e.g. linear pattern) instead of just a node

leslie-fang-intel replied:
Thanks. Added.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+-  Step2: Define the ``QuantizationSpec`` for two inputs and one output of the ``add`` node to specify
+   how to quantize input tensors and output tensor which includes parameters of ``observer type``,
+   ``dtype``, ``quant_min``, and ``quant_max`` etc.
+-  Step3: Annotate the inputs and output of the ``add`` node. User will create the ``QuantizationAnnotation``
```
I think it would be helpful to explain the structure of QuantizationAnnotation here as well, i.e. what input_qspec_map and output_qspec means, and potentially give an example here about two nodes connected together, e.g.
conv1 -> conv2, how would we annotate both to int8 and what would quant flow do in this case for the output of conv1/input of conv2

leslie-fang-intel replied:
Thanks. ``QuantizationAnnotation`` is added.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+   to match the operator pattern. 2. User may go through the nodes from start to the end and compare
+   the node's target type.
+-  Step2: Define the ``QuantizationSpec`` for two inputs and one output of the ``add`` node to specify
+   how to quantize input tensors and output tensor which includes parameters of ``observer type``,
```
I think it would be helpful to explain what ``observer type`` is, it's actually a constructor, not really a type, it can serve two purposes,
(1) allow users to pass around additional parameters on top of the fields specified in QuantizationSpec, e.g. MinMaxObserver(eps=...)
(2) allow users to have custom observer or fake quantizers

leslie-fang-intel replied:
Thanks for the suggestion. After refactor, this part has been removed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+
+::
+
+    def _annotate_add(
```
I feel we do not need to paste this whole code, the tutorial should focus on explaining every line of code instead, we can extract the relevant part and put that in each section and have good explanations for them so that users can take this and implement their own annotation

jerryzh168 replied:
maybe some of the explanations can be in the docstrings, not necessarily here, we don't need to go into a lot of details but mostly focus on explaining how these things are connected here

jerryzh168 reviewed (changes requested):
high level I think we should remove all the code examples in section 1-4 and focus on highlighting the important parts and explain how different things are connected together

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+  `get_bias_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#LL42C5-L42C19>`__
+  can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific node.
+
+.. code:: ipython3
```
please remove this code as well

leslie-fang-intel replied:
Thanks this code are removed and add a gist link here.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,630 @@
+    quantizer = qq.QNNPackQuantizer()
+    operator_config = qq.get_symmetric_quantization_config(is_per_channel=True)
+    quantizer.set_global(operator_config)
+    prepared_graph_module = prepare_pt2e_quantizer(exported_graph_module, quantizer)
```
can you add a comment here saying `prepare_pt2e_quantizer` will be updated to `prepare_pt2e` soon?

leslie-fang-intel replied:
Thanks and add the note.

leslie-fang-intel commented:
> high level I think we should remove all the code examples in section 1-4 and focus on highlighting the important parts and explain how different things are connected together

Hi @jerryzh168, addressed the comments. Please help to take a look again.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+To address these scalability issues, 
+`Quantizer <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L160>`__
+is introduced for quantization in PyTorch 2.0 export. ``Quantizer`` is a class that users can use to
+programmably set the observer or fake quant objects for each node in the model graph. It adds flexibility
```
`programmably set the observer or fake quant objects for each node in the model graph.` I think this is out of date, we should say `programmably set the quantization specifications for input and output of each node in the model graph.`

leslie-fang-intel replied:
Thanks and Fixed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+to the quantization API and allows modeling users and backend developers to configure quantization programmatically.
+This will allow users to express how they want an operator pattern to be observed in a more explicit
+way by annotating the appropriate nodes. A backend specific quantizer inherited from base quantizer,
+some APIs need to be overrided:
```
```suggestion
some methods that need to be implemented:
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+some APIs need to be overrided:
+
+-  `annotate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L269>`__
+   is used to annotate nodes in the graph with observer or fake quant constructors to convey the desired way of quantization.
```
```suggestion
   is used to annotate nodes in the graph with QuantizationAnnotation objects to convey the desired way of quantization.
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+
+-  `annotate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L269>`__
+   is used to annotate nodes in the graph with observer or fake quant constructors to convey the desired way of quantization.
+-  `validate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L721>`__
```
I think maybe we can skip this for now? I think we haven't figure out the detail for this one yet?

cc @kimishpatel what do you think?

leslie-fang-intel replied:
Thanks and temporarily removed the ``validate method``.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+
+-  Step 2: Define the ``QuantizationSpec`` for inputs and output of the pattern. ``QuantizationSpec``
+   defines the ``data type``, ``qscheme``, and other quantization parameters about users' intent of
+   how to observer/quantize a tensor.
```
```suggestion
   how to observe or fake quantize a tensor.
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+can be an input edge or an output value. 
+
+-  Input edge is the connection between input node and the node consuming the input,
+   so it's a Tuple[Node, Node].
```
```suggestion
   so it's a ``Tuple[Node, Node]``.
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+
+-  Input edge is the connection between input node and the node consuming the input,
+   so it's a Tuple[Node, Node].
+-  Output value is an fx Node.
```
```suggestion
-  Output value is an fx ``Node``.
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+two input tensors as sharing quantization parameters. We can define its ``QuantizationAnnotation``
+as this:
+
+-  Step 1: Annotate input_act0 of ``add`` with ``QuantizationSpec``.
+-  Step 2: Create a ``SharedQuantizationSpec`` object with input edge defined as ``(input_act0, add_node)`` which means to
+   share the observer used for this edge. Then, user can annotate input_act1 with this ``SharedQuantizationSpec``
+   object.
```
nit: step 1 should be find the pattern, the same as the other examples, could you add that? same for section 3 and section 4

leslie-fang-intel replied:
Thanks. Add the ``pattern match`` as step1 for both section 2, 3, 4.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+        quant_min=0,
+        quant_max=255,
+        qscheme=torch.per_tensor_affine,
+        scale=2.0 / 256.0,
+        zero_point=128,
```
```suggestion
        scale=1.0 / 256.0,
        zero_point=0,
```

I remember these have been updated

leslie-fang-intel replied:
Oh, yean. Thanks and updated.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+-  Step 1: Define ``derive_qparams_fn`` function, it accepts list of ``ObserverOrFakeQuantize`` (
+   `ObserverBase <https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/ao/quantization/observer.py#L124>`__
+   or `FakeQuantizeBase <https://github.com/pytorch/pytorch/blob/07104ca99c9d297975270fb58fda786e60b49b38/torch/ao/quantization/fake_quantize.py#L60>`__)
+   as input. From each ``ObserverOrFakeQuantize`` object, user can get the ``scale``, ``zero point`` value. Combine these values together,
+   user can define its heuristic about how to derive new ``scale``, ``zero point`` value.
```
```suggestion
   as input. From each ``ObserverOrFakeQuantize`` object, user can get the ``scale``, ``zero point`` value. User can define its heuristic about how to derive new ``scale``, ``zero point`` value based on the quantization parameters calculated from the observer or fake quant instances.
```

leslie-fang-intel replied:
Thanks and update the description.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+
+After above annotation methods defined with ``QuantizationAnnotation API``, we can now put them together to construct a ``BackendQuantizer``
+and run a `toy example <https://gist.github.com/leslie-fang-intel/b78ed682aa9b54d2608285c5a4897cfc>`__
+with ``Torchvision Resnet18``. To better understand the final example, here are some basic concepts which are used:
```
```suggestion
with ``Torchvision Resnet18``. To better understand the final example, here are the classes and utility functions that are used in the example:
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+
+- `QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
+  consists of ``QuantizationSpec`` for activation, weight, and bias separately.
+- When annotating the model, methods of
```
```suggestion
- When annotating the model,
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,341 @@
+- `QuantizationConfig <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L103-L109>`__
+  consists of ``QuantizationSpec`` for activation, weight, and bias separately.
+- When annotating the model, methods of
+  `get_act_qspec <https://github.com/pytorch/pytorch/blob/73fd7235ad25ff061c087fa4bafc6e8df4d9c299/torch/ao/quantization/_pt2e/quantizer/utils.py#L9>`__,
```
nit: this is now changed to https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/utils.py#L10 and https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/_pt2e/quantizer/utils.py#L23

leslie-fang-intel replied:
Thanks and the link are updated.

leslie-fang-intel commented:
Hi @jerryzh168, the comments has been addressed. Please help to take a look again. Thanks.

jerryzh168 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+   `get_output_act_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L23>`__,
+   `get_weight_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L36>`__, and
+   `get_bias_qspec <https://github.com/pytorch/pytorch/blob/47cfcf566ab76573452787335f10c9ca185752dc/torch/ao/quantization/_pt2e/quantizer/utils.py#L53>`__
+   can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific node.
```
```suggestion
   can be used to get the ``QuantizationSpec`` from ``QuantizationConfig`` for a specific pattern.
```

leslie-fang-intel replied:
Thanks and changed.

jerryzh168 reviewed (approved):
LGTM, thanks!

jerryzh168 commented:
@kimishpatel please take a look again as well

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+(prototype) Quantization in PyTorch 2.0 Export Tutorial
```
Can we slap work in progress for this?

leslie-fang-intel replied:
Thanks. Add  `Work in Progress` in the title.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+
+Previously in ``FX Graph Mode Quantization`` we were using ``QConfigMapping`` for users to specify how the model to be quantized
+and ``BackendConfig`` to specify the supported ways of quantization in their backend.
+This API covers most use cases relatively well, but the main problem is that this API is not fully extensible
```
@jerryzh168 changes to new API is not so much about whether it covers the cases well or not, it is just a new way of thinking of the quantizaiton workflow in the context of PT2 export. So I dont think it is quite accurate to say that we built new workflow to cover tiny percentage of cases that werent working well.

THis is also the reason why i dont want to jump into documenting something that is not quite true to the reason of why new API exist. cc: @supriyar

jerryzh168 replied:
I feel maybe we should mention (3) and (4) as well in https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c which are not limitations but more talking about motivations for the new API to have the complete picture?

leslie-fang-intel replied:
Thanks for the suggestions. section 3, 4 are added.

supriyar replied:
I don't think we should be documenting the details on the motivation for quantizer and annotation API to this level in this tutorial. Summarizing the motivation in 1-2 sentences should be good and the details can be left to the appendix or some other detailed doumentation/notes.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+-  Limitation around expressing quantization intentions for complicated operator patterns such as in the discussion of
+   `issue-96288 <https://github.com/pytorch/pytorch/issues/96288>`__ to support ``conv add`` fusion with oneDNN library.
+   It also requires some changes to current already complicated pattern matching code such as in the
+   `PR-97122 <https://github.com/pytorch/pytorch/pull/97122>`__ to support ``conv add`` fusion.
```
This is an outcome of the new API, not necessarily the intention with which it was developed. So to say that these are two limitations for which we are developing new API is not quite accurate.

leslie-fang-intel replied:
Thanks. I think it's similar as previous comment. So as suggested by Jerry in here https://github.com/pytorch/tutorials/pull/2354#discussion_r1223618560. 2 more bullets are added to explain the motivations.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+
+    # Step 4: Lower Reference Quantized Model into the backend
+
+Inside the Quantizer, we will use the ``QuantizationAnnotation API``
+to convey user's intent for what quantization spec to use and how to
+observe certain tensor values in the prepare step. Now, we will have a step-by-step
+tutorial for how to use the ``QuantizationAnnotation API`` with different types of
+``QuantizationSpec``.
```
Befor we introduce any of this, we need to explain what annotation API does. For that we first also have to explain quantization spec.

So the flow I imagine would be
```
Quantizer uses annotation API to convey quantization intent for different operators/patterns. Annotation API uses QuantizationSpec to convey intent of how a tensor will be quantized, e.g. dtype, bitwidth, min, max values, symmetric vs. asymmetric etc. See the spec definition at.... Furthermore it also allows quantizer to specify how a tensor value should be observed, e.g. minmax vs. histogram vs. something custom.

Quantization spec is then used to annotate nodes' output tensor or input tensors. Annotating input tensors is equivalent of annotating edge of the graph, while annotating output tensor is equivalent of annotating node. Thus annotation API requires quantizer to annotate nodes (output tensor) or edges (input tensors) of the graph.

For example...
```

leslie-fang-intel replied:
Thanks for the suggestion. It makes the introduction of ``Quantizer`` and ``QuantizationSpec`` much clearer.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+
+In order to use the quantized pattern/operators, e.g. ``quantized add``,
+backend developers will have intent to quantize (as expressed by
+`QuantizationSpec <https://github.com/pytorch/pytorch/blob/1ca2e993af6fa6934fca35da6970308ce227ddc7/torch/ao/quantization/_pt2e/quantizer/quantizer.py#L38>`__
```
We need to explain what is quantization spec first

leslie-fang-intel replied:
Thanks. I think it's same as above, so add the explanation of ``QuantizationSpec`` there.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+of how this intent is conveyed in the quantization workflow with annotation API.
+
+-  Step 1: Identify the original floating point pattern in the FX graph. There are
+   several ways to identify this pattern: User may use a pattern matcher (e.g. SubgraphMatcher)
```
`User` => `Quantizer`

leslie-fang-intel replied:
Thanks and changed.

kimishpatel commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,350 @@
+-  Step 1: Identify the original floating point pattern in the FX graph. We can use the same
+   methods introduced in ``QuantizationSpec`` example to identify the ``add`` pattern.
+-  Step 2: Annotate input_act0 of ``add`` with ``QuantizationSpec``.
+-  Step 3: Create a ``SharedQuantizationSpec`` object with input edge defined as ``(input_act0, add_node)`` which means to
```
THis is exactly why we need to preface this by saying that annotation API requires quantizer to annotate nodes or edges. Nodes are annotated via `output_qspec` and edges can be annotated via `input_qspec_map`. Later we will show in edge can also be explicitly annotated when quantizer wants to express intention of sharing quantization parameters or deriving them.

kimishpatel reviewed (approved):
I left several comments. I feel overall this is not painting the picture of why we are doing new API and thus the motivaiton. HOwever, since I am reviewing this too late, I dont want to block this on my behalf.

leslie-fang-intel commented:
Hi @jerryzh168 @kimishpatel, thanks for the suggestions. The comments has been addressed. Please help to take a look again.

jgong5 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,372 @@
+   that is incompatible with a specific ``BackendConfig``. This is not a great UX. Ideally, we can structure this better
+   by making both configuration (``QConfigMapping``) and quantization capability (``BackendConfig``) backend
+   specific. So there will be less confusion about incompatibilities.
+-  Currently, in ``QConfig`` we are exposing observer/fake_quant classes as an object for user to configure quantization.
+   This increases the things that user needs to care about, e.g. not only the ``dtype`` but also how the observation should
+   happen. These could potentially be hidden from user to make user interface simpler.
```
A bit confused by this, I think the new API also has the observer setting in the `QuantizationSpec`. So it is also part of the new API, right?

leslie-fang-intel replied:
Thanks for the comment. In creation of `QuantizationSpec`, we do still need to specify the class of `observer`. But I think it's simpler for users to use comparing with  `FX Quantization Mode`.

- In `FX Quantization Mode`, `QConfig` is created with observer. User needs to learn the constructor and specify the quantization parameters for different observers in order to create a `QConfig` for his use case.
- Now the general quantization parameters such as ``dtype``, ``qscheme`` for different observers are unified in the ``QuantizationSpec`` and only the observer class are needed.

@jerryzh168 I think you may help to comment more about this bullet.

jerryzh168 replied:
so in the new quantizer API, the idea is that configurability is controlled by each backend, the quantizer/annotation API is not facing common modeling user directly, it is an API for backend developers or advanced modeling users only. for common users, they will be interacting with each backend specific quantizer only, e.g. I can have a quantizer that only exposes "quantize/not quantize" option to users:

```
backend_quantizer = BackendQuantizer()
backend_quantizer.enable_quantization()
model = prepare_pt2e(model, backend_quantizer)
...
```

see the graph in the end of motivation section: https://docs.google.com/document/d/1_jjXrdaPbkmy7Fzmo35-r1GnNKL7anYoAnqozjyY-XI/edit#heading=h.jtqauapwj95c for details

jgong5 commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,372 @@
+``QuantizationSpec`` is used to annotate nodes' input tensors or output tensor. Annotating
+input tensors is equivalent of annotating edge of the graph, while annotating output tensor is
+equivalent of annotating node. Thus annotation API requires quantizer to annotate
+edges (input tensors) or nodes (output tensor) of the graph.
```
I think it is also worth noting here about `QuantizationAnnotation` and its relationship with `QuantizationSpec`. If we view the model as a graph,  `QuantizationAnnotation` is for node (op) and `QuantizationSpec` is for edge (tensor). `QuantizationAnnotation` is composed of the `QuantizationSpec` for all the input and output tensor of an op.

jgong5 reviewed (approved):
LGTM.

supriyar commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,375 @@
+Prerequisites:
+-----------------------
+
+-  `Understanding of torchdynamo concepts in PyTorch <https://pytorch.org/docs/stable/dynamo/index.html>`__
+-  `Understanding of the quantization concepts in PyTorch <https://pytorch.org/docs/master/quantization.html#quantization-api-summary>`__
+-  `Understanding of FX Graph Mode post training static quantization <https://pytorch.org/tutorials/prototype/fx_graph_mode_ptq_static.html>`__
+-  `Understanding of BackendConfig in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html?highlight=backend>`__
+-  `Understanding of QConfig and QConfigMapping in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html#set-up-qconfigmapping-that-satisfies-the-backend-constraints>`__
+
```
nit: you can remove "Understanding of" from all these bullet points and keep them a bit more concise.

leslie-fang-intel replied:
Thanks for comment and removed.

supriyar commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,375 @@
+-  `Understanding of BackendConfig in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html?highlight=backend>`__
+-  `Understanding of QConfig and QConfigMapping in PyTorch Quantization FX Graph Mode <https://pytorch.org/tutorials/prototype/backend_config_tutorial.html#set-up-qconfigmapping-that-satisfies-the-backend-constraints>`__
+
+Previously in ``FX Graph Mode Quantization`` we were using ``QConfigMapping`` for users to specify how the model to be quantized
```
Is this a separate section? Currently it seems to fall under Prerequisites

leslie-fang-intel replied:
Thanks, I will add it ``Preface`` section.

supriyar commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,375 @@
+way by annotating the appropriate nodes. A backend specific quantizer inherited from base quantizer,
+some methods that need to be implemented:
+
+-  `annotate method <https://github.com/pytorch/pytorch/blob/3e988316b5976df560c51c998303f56a234a6a1f/torch/ao/quantization/_pt2e/quantizer/qnnpack_quantizer.py#L269>`__
```
is this the only method that needs to be implemented (above description says some methods)? Why create a separate bullet point for it in that case?

leslie-fang-intel replied:
Actually there are some other methods needs to be implemented. However, the detail design for the other methods are not fully set per the discussion with @jerryzh168. So we only mention `annotate method ` here. Sure, I will merge this bullet into previous paragraph for now.

supriyar commented on prototype_source/quantization_in_pytorch_2_0_export_tutorial.rst:
```diff
@@ -0,0 +1,375 @@
+nodes (output tensor) of the graph. Now, we will have a step-by-step tutorial for
+how to use the annotation API with different types of ``QuantizationSpec``.
+
+1. Annotate common operator patterns
```
these seem to be individual headings. I'd recommend having a heading on Annotation API Usage and then include these within that for better formatting.

leslie-fang-intel replied:
Thanks and changed.

leslie-fang-intel commented:
@supriyar Thanks for the comments. I made another PR https://github.com/pytorch/tutorials/pull/2456 to following up your comments. Please help to take a look.
